### PR TITLE
Implement SwappableEventBase

### PR DESCRIPTION
**Pull request by dymk**
Implements SwappableEventBase for use in RSocketRequester,
to allow swapping of EventBases while maintaining order of enqueued
tasks

Test Plan: `make; ./tests`

### CODE REVIEW

File: test/internal/SwappableEventBaseTest.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  DidExecTracker(int line, std::string file, std::string name)
+    : line(line), file(file), name(name) {}
+  MOCK_METHOD0(mark, void());
+};
+
+struct DETMarkedOnce : public ::testing::CardinalityInterface {
+  explicit DETMarkedOnce(DidExecTracker const& det) : det(det) {}
+  DidExecTracker const& det;
+
+  virtual int ConservativeLowerBound() const override { return 1; }
```
Comment (lehecka) [inline test/internal/SwappableEventBaseTest.cpp]: nit: in fb the coding style recommends avoiding repeating virtual keyword when override is present.

File: src/internal/SwappableEventBase.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+// SwappableEventBase provides an interface similar to EventBase, allowing
+// an underlying EventBase to be changed, and to force callbacks to be
+// executed in serial order regardless of which underlying EventBase they are
+// enqueued on.
+class SwappableEventBase final {
+  // std::mutex doesn't like being in a std::pair
+  struct MutexBoolPair {
+    // lock for synchronization on destroyed_, and all members of the parent SEB
+    std::mutex l_;
```
Comment (lehecka) [inline src/internal/SwappableEventBase.h]: nit: public fields should not have _ suffix

**Comment by dymk** [comment]
I'm unable to reproduce the crashes seen by Travis; flakey tests?

**Comment by alexmalyshev** [comment]
I think I fixed them with https://github.com/rsocket/rsocket-cpp/pull/604

**Pull request merged**